### ISSUE TITLE
[lsan] Trivial leak missed on x86

### ISSUE THREAD

**Issue by stevenvar** [google/sanitizers#1447]
When I run lsan on the following code: 

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
 
int main(int argc, char** argv){
  char* s = (char *)malloc(3);
  strcpy(s,"aa");
  printf("%s\n",s);
  return 0;
}
```

After using the following compiling command:

```
clang test.cpp -fsanitize=address -g -m32
```

The leak sanitizer misses the obvious leak from the s pointer. Weirdly enough using -fsanitize=address instead of leak and using any optimization level different than -O0 fixes the issue and the leak is detected...

Is this "normal" behaviour? Does anyone know for what reason that would happen?

When running in 64 bits mode everything is working OK.

Thanks.

**Comment by yugr** [comment]
Yes, Lsan is not guaranteed to detect all leaks, see e.g. #937 for some explanations.
